app/src/main/java/com/hedvig/app/feature/genericauth/otpinput/OtpInputViewModel.kt

```diff
@@ -0,0 +1,111 @@
+package com.hedvig.app.feature.genericauth.otpinput
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.hedvig.app.authenticate.AuthenticationTokenService
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.receiveAsFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+
+class OtpInputViewModel(
+    private val otpId: String,
+    private val credential: String,
+    private val authenticationTokenService: AuthenticationTokenService,
+    private val sendOtpCodeUseCase: SendOtpCodeUseCase,
+    private val reSendOtpCodeUseCase: ReSendOtpCodeUseCase,
+) : ViewModel() {
+    private val _viewState = MutableStateFlow(ViewState(credential = credential))
+    val viewState = _viewState.asStateFlow()
+
+    private val _events = Channel<Event>(Channel.UNLIMITED)
+    val events = _events.receiveAsFlow()
```

Comment:
I'm slightly scared that this is incorrect and will drop events, which is embarrasing as I was the one who put this pattern in place in the first place.

It appears to be really hard to figure out what the truly correct method is, but I think the correct code is probably the following:

```suggestion
    val events: ReceiveChannel = _events
```

this could then be simplified to the following:

```kotlin
fun <T> Channel<T>.asReceiver(): ReceiveChannel<T> = this
```

with the consumer being converted to:

```kotlin
model
	.events
	.consumeAsFlow()
	.flowWithLifecycle(lifecycleOwner)
	.onEach { /* ... */ }
	.launchIn(lifecycleScope)
```

If my fear is correct, then we risk the same data loss that we got when using `SharedFlow` for events happening here as well, which would not be ideal.

We don't need to remedy this issue in the context of this PR, but I could take it upon myself to refactor all uses of the current pattern to what I believe the be the correct method.

Check my code before we commit to that, though! @hugokallstrom @StylianosGakis

Reply:
That sounds reasonable imo!